leanplum_data_export/export_streaming.py

```diff
@@ -0,0 +1,172 @@
+
+import boto3
+
+from .base_exporter import BaseLeanplumExporter
+
+
+class StreamingLeanplumExporter(BaseLeanplumExporter):
+
+    DATA_TYPES = [
+        "eventparameters", "events", "experiments", "sessions", "states", "userattributes"
+    ]
+    FILE_HISTORY_PREFIX = "file_history"
+
+    def __init__(self, project):
+        super().__init__(project)
+        self.s3_client = boto3.client("s3")
+
+    def get_files(self, date: str, bucket: str, prefix: str, max_keys: int = None) -> List[str]:
+        """
+        Get the s3 keys of the data files in the given bucket
+        """
+        max_keys = {} if max_keys is None else {"MaxKeys": max_keys}  # for testing pagination
+        filename_re = re.compile(r"^.*/\d{8}/export-.*-output-([0-9]+)$")
+        data_file_keys = []
+
+        continuation_token = {}  # value used for pagination
+        while True:
+            object_list = self.s3_client.list_objects_v2(
+                Bucket=bucket,
+                Prefix=os.path.join(prefix, date, "export-"),
+                **continuation_token,
+                **max_keys,
+            )
+            data_file_keys.extend([content["Key"] for content in object_list["Contents"]
+                                   if filename_re.fullmatch(content["Key"])])
+
+            if not object_list["IsTruncated"]:
+                break
+
+            continuation_token["ContinuationToken"] = object_list["NextContinuationToken"]
+
+        return data_file_keys
+
+    def get_previously_imported_files(self, bucket: str, prefix: str,
+                                      version: str, date: str) -> Set[str]:
+        """
+        Get file names of data files that have already been imported into GCS
+        """
+        blobs = self.gcs_client.list_blobs(
+            bucket,
+            prefix=self.get_gcs_prefix(prefix, version, date, self.FILE_HISTORY_PREFIX)
+        )
+        file_names = []
+        for page in blobs.pages:
+            for blob in page:
+                file_names.append(os.path.basename(blob.name))
+
+        return set(file_names)
+
+    def write_to_gcs(self, file_path: Union[Path, str], data_type: str, bucket: str,
+                     prefix: str, version: str, date: str) -> None:
+        """
+        Write file to GCS bucket
+        If a Path is given as file_path, the file is uploaded
+        If a string is given as file_path, an empty file is uploaded
+        """
+        file_name = file_path if isinstance(file_path, str) else file_path.name
+        gcs_path = os.path.join(self.get_gcs_prefix(prefix, version, date, data_type), file_name)
+
+        logging.info(f"Uploading {file_name} to gs://{gcs_path}")
+        blob = self.gcs_client.bucket(bucket).blob(gcs_path)
+        if isinstance(file_path, str):
+            blob.upload_from_string("")
+        else:
+            blob.upload_from_filename(str(file_path))
+
+    def write_to_csv(self, csv_writers: Dict[str, csv.DictWriter], session_data: Dict,
```

Comment:
Hmm hmm hmm.

So the JSON data has one big blob that contains everything about the session, while the previous CSV import separated those out into tables.

What are your thoughts on using the JSON format itself? While this does mean the new tables are schema incompatible with the old, it keeps all session data together and simplifies the import a great deal.

Reply:
I tried loading the json directly at first but the format of the json isn't loadable by bigquery since some keys have spaces in them which makes them invalid column names.  Also for fields like user attributes the key is the attribute name which would make the schema dynamic which complicates things.  So either way we would have to transform the file before loading it and I figured it would be easiest if I just copied the existing way.

Reply:
Right, that makes sense; I remember that when I first did the import. Carry on then.